src/System.Private.CoreLib/shared/System/MemoryExtensions.Trim.cs

```diff
@@ -0,0 +1,818 @@
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            int length = ClampEnd(span, start, trimElement);
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int length = ClampEnd(span, 0, trimElement);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampStart<T>(ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = 0;
+
+            if (trimElement == null)
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (span[start] != null)
+                        break;
+                }
+            }
+            else
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (!trimElement.Equals(span[start]))
+                        break;
+                }
+            }
+
+            return start;
+        }
+
+        /// <summary>
+        /// Delimits all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="start">The start index from which to being searching.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampEnd<T>(ReadOnlySpan<T> span, int start, T trimElement)
+            where T : IEquatable<T>
+        {
+            // Initially, start==len==0. If ClampStart trims all, start==len
+            Debug.Assert((uint)start <= span.Length);
```

Comment:
Why the cast to `uint`? Should we also assert `start >= 0`?

Reply:
It's a trick to test two boundaries with one statement:
If the caller passed in `-1` then `(uint)-1 <= len` is `false` (since `4294967295 > len`)
If they passed in `0` then `0u <= len` is `true`
If they passed in `100` then `100u <= len` depends on `len`

Reply:
Right, of course. I generally haven't seen the optimization used within `Debug.Asserts` since it doesn't matter if we have extra branches there.

Reply:
I have used this trick everywhere since I learnt it, so it's habit now. You're right it's not strictly necessary here